book/src/dev/rfcs/0002-parallel-verification.md

```diff
@@ -0,0 +1,335 @@
+  are out of scope for this RFC.
+
+This RFC focuses on Semantic Verification, and the design patterns that enable
+blocks to be verified in parallel.
+
+## Verification Interfaces
+[verification-interfaces]: #verification-interfaces
+
+Verification is implemented by the following traits and services:
+* **Structural Verification:**
+  * `zebra_network::init`: Provides a downloader service that accepts a
+    `BlockHeaderHash` request, and parses the peer response into a `Block`
+    struct.
+  * `zebra_chain::ZcashDeserialize`: A trait for parsing consensus-critical
+    data structures from a byte buffer.
+* **Semantic Verification:**
+  * `ChainVerifier`: Provides a verifier service that accepts a `Block` request,
+    performs verification on the block, and responds with a `BlockHeaderHash` on
+    success.
+  * Internally, the `ChainVerifier` selects between a `CheckpointVerifier` for
+    blocks that are within the checkpoint range, and a `BlockVerifier` for
+    recent blocks.
+* **State Updates:**
+  * `zebra_state::init`: Provides the state update service, which accepts
+    requests to add blocks to the state.
+
+### Checkpoint Verification
+[checkpoint-verification]: #checkpoint-verification
+
+The `CheckpointVerifier` performs rapid verification of blocks, based on a set
+of hard-coded checkpoints. Each checkpoint hash can be used to verify all the
+
+previous blocks, back to the genesis block. So Zebra can skip almost all
+verification for blocks in the checkpoint range.
+
+The `CheckpointVerifier` uses an internal queue to store pending blocks.
+Checkpoint verification is cheap, so it is implemented using non-async
+functions within the CheckpointVerifier service.
+
+Here is how the `CheckpointVerifier` implements each verification stage:
+
+* **Structural Verification:**
+  * *As Above:* the `CheckpointVerifier` accepts parsed `Block` structs.
+* **Semantic Verification:**
+  * `check_height`: makes sure the block height is within the unverified
+    checkpoint range, and adds the block to its internal queue.
+  * `target_checkpoint_height`: Checks for a continuous range of blocks from
+    the previous checkpoint to a subsequent checkpoint. If the chain is
+    incomplete, returns a future, and waits for more blocks. If the chain is
+    complete, assumes that the `previous_block_hash` fields of these blocks
+    form an unbroken chain from checkpoint to checkpoint, and starts
+    processing the checkpoint range. (This constraint is an implicit part of
+    the `CheckpointVerifier` design.)
+  * `process_checkpoint_range`: makes sure that the blocks in the checkpoint
+    range have an unbroken chain of previous block hashes.
+* **State Updates:**
+  * *As Above:* the `CheckpointVerifier` returns success to the `ChainVerifier`,
+    which sends verified `Block`s to the state service.
+  * Zebra implements the block height consensus rule, which makes sure
+    each block is added to state after its previous block.
```

Comment:
```suggestion
  * *As Above:* the `CheckpointVerifier` returns success to the `ChainVerifier`,
    which sends verified `Block`s to the state service to be committed.
```
I think this contradicts the example above of block height checks, where the state has to buffer queued contextual verification requests, but I don't think we lose much if we omit the last point.

Also, I think that we will want to have two different kinds of state request, one for use by the checkpoint verifier and one for use by the block verifier, since we can skip all the contextual checks.  I'm not sure how that should be described in this section, but since it's really something about the state layer, which is out of scope for this RFC, I'm not sure it's important to do here.

Reply:
I think optimisations and the state layer are both out of scope for this RFC.

Reply:
Fixed in fb53bb8b7f45e8659b0e1f2d35ad8e01f1577b5c.